trinity/sync/full/state.py

```diff
@@ -253,11 +230,42 @@ def deregister_peer(self, peer: BasePeer) -> None:
                 return
 
             candidates = list(not_yet_requested.difference(self._peer_missing_nodes[peer]))
-            batch = candidates[:eth_constants.MAX_STATE_FETCH]
+            batch = tuple(candidates[:eth_constants.MAX_STATE_FETCH])
             not_yet_requested = not_yet_requested.difference(batch)
             self.request_tracker.active_requests[peer] = (time.time(), batch)
-            self.logger.debug("Requesting %d trie nodes to %s", len(batch), peer)
-            peer.sub_proto.send_get_node_data(batch)
+            asyncio.ensure_future(self._request_and_process_nodes(peer, batch))
+
+    async def _request_and_process_nodes(self, peer: ETHPeer, batch: Tuple[Hash32, ...]) -> None:
+        self.logger.debug("Requesting %d trie nodes from %s", len(batch), peer)
+        nodes = await peer.requests.get_node_data(batch)
+
+        self.logger.debug("Got %d NodeData entries from %s", len(nodes), peer)
+        _, requested_node_keys = self.request_tracker.active_requests.pop(peer)
+
+        loop = asyncio.get_event_loop()
+        node_keys = await loop.run_in_executor(self._executor, list, map(keccak, nodes))
+
+        # check for missing nodes and re-schedule them
+        missing = set(batch).difference(node_keys)
+
+        self._peer_missing_nodes[peer].update(missing)
+        if missing:
+            await self.request_nodes(missing)
+
+        # check for unexpected nodes and log if there are any
+        unexpected = set(node_keys).difference(batch)
+        if unexpected:
```

Comment:
Can this actually happen or would a ValidationError be raised by the `get_node_data()` call above if a node sent us unexpected nodes?